[libclc] Initial support for cross-compiling OpenCL libraries

jhuber6 opened the pull request:
Summary:
The other GPU enabled libraries, (openmp, flang-rt, compiler-rt, libc,
libcxx, libcxx-abi) all support builds through a runtime cross-build. In
these builds we use a separate CMake build that cross-compiles to a
single target.

This patch provides basic support for this with the `libclc` libraries.
Changes include adding support for the more standard GPU compute triples
(amdgcn-amd-amdhsa, nvptx64-nvidia-cuda) and building only one target in
this mode.

Some things left to do:

This patch does not change the compiler invocations, this method would
allow us to use standard CMake routines but this keeps it minimal.

The prebuild support is questionable and doesn't fit into this scheme
because it's a host executable, I'm ignoring it for now.

The installed location should just use the triple with no `libclc/`
subdirectory handling I believe.

arsenm reviewed (commented):
I don't think cmake provides out of the box OpenCL support as a compile language. At one point we had the boilerplate to add it as a language here

jhuber6 commented on libclc/cmake/modules/AddLibclc.cmake:
```diff
@@ -454,10 +454,22 @@ function(add_libclc_builtin_set)
-    )
+    if( TARGET prepare_builtins )
+      # FIXME: Is this utility even necessary? The `-mlink-builtin-bitcode`
+      # option used to link the library in discards the modified linkage.
```
The two things it does from what I can tell

1. Removes `"opencl.ocl.version"` metadata strings

2. Replaces any non-external function with `linkonce_odr` linkage

The first one, I'm pretty sure `llvm-link` deduplicates now. For the second one, this is mostly just a cheap way to let the functions link while being eliminated by the compiler if unused. I think the ROCm Device Libs do that as well

jhuber6 commented:
@wenju-he @frasercrmck General question, how difficult would it be to port this project to use `add_library` instead? We would need some custom language support, that's just CMake boilerplate. Getting the final linked library can easily be done with a custom command to call `llvm-link` as `llvm-link` on a static library will link everything inside it.

jhuber6 commented on libclc/CMakeLists.txt:
```diff
@@ -82,7 +87,7 @@ else()
   set( LIBCLC_STANDALONE_BUILD FALSE )
 
-  set( LLVM_PACKAGE_VERSION ${LLVM_VERSION} )
+  set( PACKAGE_VERSION ${LLVM_PACKAGE_VERSION} )
```
I'm a little confused what `in-tree` means here. If it's in the context like we're doing here with a runtimes build it should be set.

wenju-he replied:
> I'm a little confused what `in-tree` means here. If it's in the context like we're doing here with a runtimes build it should be set.

sorry, you're right that `LLVM_PACKAGE_VERSION` is set in in-tree build when libclc is in `LLVM_ENABLE_RUNTIMES`.

It is just that our downstream still uses `LLVM_ENABLE_PROJECTS="...,libclc"` and in this case `LLVM_PACKAGE_VERSION` is not set. The reasons of not switching to `LLVM_ENABLE_RUNTIMES` yet are:
* prepare_builtins has build fail when libclc is in LLVM_ENABLE_RUNTIMES on windows when MSVC generator is used, see https://github.com/google/clspv/pull/1521#issuecomment-3287406308 and analysis at https://github.com/google/clspv/pull/1521#issuecomment-3350179501. CMAKE_C_COMPILER set at https://github.com/llvm/llvm-project/blob/80b62cb9b979bf560ea17c127b1e85749376a6a2/llvm/cmake/modules/LLVMExternalProjectUtils.cmake#L228 doesn't work. This could be an blocking issue for switching libclc to `add_library` since CMAKE_C_COMPILER  will be used for compiling .cl files.
* `LLVM_ENABLE_RUNTIMES="libclc"` build is much slower than `LLVM_ENABLE_PROJECTS="libclc"` in debug build when compiling a execution-time support library which depends on libclc. I have a local workaround for this issue and long term solution might be removing the dependency on libclc.

jhuber6 replied:
Alright, so just setting both for now is the easiest?

wenju-he commented on libclc/CMakeLists.txt:
```diff
@@ -206,10 +218,12 @@ set( r600--_devices cedar cypress barts cayman )
 set( amdgcn--_devices tahiti )
 set( amdgcn-mesa-mesa3d_devices ${amdgcn--_devices} )
 set( amdgcn--amdhsa_devices none )
+set( amdgcn-amd-amdhsa_devices none )
```
perhaps we can drop amdgcn--amdhsa_devices target from libclc since the bitcode will be the same?

jhuber6 replied:
I was planning on trimming this up in a PR tomorrow. We definitely could though I don't know the subtleties here, since maybe someone depends on the specific triple.

I also really want to change how the files are installed, we should use the regular triple and then use the specific device as a subdirectory if possible. Right now we have like `libclc/` in the resource dir, when it should be more like standard `PER_TARGET_RUNTIME_DIR` I'd say.

arsenm replied:
The amd vendor only does anything with spirv

wenju-he commented on libclc/CMakeLists.txt:
```diff
@@ -206,10 +218,12 @@ set( r600--_devices cedar cypress barts cayman )
 set( clspv64--_devices none )
 set( nvptx--_devices none )
 set( nvptx64--_devices none )
+set( nvptx64-nvidia-cuda_devices none )
```
I don't know about history of using nvidiacl in libclc. If nvptx64-nvidia-cuda is the right choice, we can probably rename ptx-nvidiacl folders so that nvptx64-nvidia-cuda will pick up files in those folders and drop nvptx64--nvidiacl target.

jhuber6 committed:
[libclc] Initial support for cross-compiling OpenCL libraries

Summary:
The other GPU enabled libraries, (openmp, flang-rt, compiler-rt, libc,
libcxx, libcxx-abi) all support builds through a runtime cross-build. In
these builds we use a separate CMake build that cross-compiles to a
single target.

This patch provides basic support for this with the libclc libraries.
Changes include adding support for the more standard GPU compute triples
(amdgcn-amd-amdhsa, nvptx64-nvidia-cuda) and building only one target in
this mode.

Some things left to do:

This patch does not change the compiler invocations, this method would
allow us to use standard CMake routines but this keeps it minimal.

The prebuild support is questionable and doesn't fit into this scheme
because it's a host executable, I'm ignoring it for now.

The installed location should just use the triple with no libclc/
subdirectory handling I believe.

jhuber6 commented:
This is much simpler now that I landed some of the other PRs, includes https://github.com/llvm/llvm-project/pull/174611

wenju-he reviewed (approved):
LGTM
